### DISCUSSION TITLE
useQuerySelector instead of `select` option in useQuery

### DISCUSSION THREAD

**Discussion by zaverden** [TanStack/query#3353] [Ideas] [answered]
from this comment for v4: https://github.com/tannerlinsley/react-query/issues/1639#issuecomment-762383500
> Remove the select feature and document how to "select" in userland with memoization.

While I aggree that the selection feature should be out of core, I think it is a good idea to provide it as an extra hook in react part of the react-query.

I propose to add this hook:
```typescript
function useQuerySelector<TData = unknown, TError = unknown, TTargetData = unknown>(

  queryResult: UseQueryResult<TData, TError>,
  selector: (v: TData) => TTargetData,
  deps: unknown[] = []

): UseQueryResult<TTargetData, TError>
```

Usage will look like this (inspired by example in the issue https://github.com/tannerlinsley/react-query/issues/3065): 
```typescript
function useTodosQuery() {
  return useQuery("todos/all", loadTodos);
}

function useTodosCountQuery() {
  const todosQuery = useTodosQuery();
  return useQuerySelector(todosQuery, (todos) => todos.length);
}

function useTodoQuery(id) {
  const todosQuery = useTodosQuery();
  return useQuerySelector(todosQuery, (todos) => todos.find((t) => t.id === id), [id]);
}
```

In my opinion, it makes it easier to reuse base query. with `select` option `useTodosQuery` had to support of being requested with different selectors (see original examples [here](https://tkdodo.eu/blog/react-query-data-transformations)). With `useQuerySelector` hook it is cleaner.

Thankfully to my poor googling skills (I did not found the `select` option when I needed it), I built `useQuerySelector` for one of my projects. if you believe it'll be a good fit into `react-query` I can  work on PR adding the hook, tests and documentation.

https://codesandbox.io/s/usequeryselector-6divd1?file=/src/use-query-selector.ts

In addition: I've been talking only about `select` from `useQuery`. But if there is a decision "v4 removes `select` option" we should cover infinite queries and (probably)mutations. It'll be much easier to handle https://github.com/tannerlinsley/react-query/issues/3065 in `useInfiniteQuerySelector` instead of messing with core types. And `useMutationSelector` will handle https://github.com/tannerlinsley/react-query/discussions/3059

**Comment by TkDodo** [comment, marked as answer]
We've iterated over selectors a lot of times since that comment, and there are currently no plans to remove them for v4. The issues mentioned in #1639 have been fixed. We now run select every time `data` changes or the referential identity to select itself changes, which might be on every render if it's an inline function (which likely doesn't matter much), but it can be optimized via  `useCallback` or by extracting it to a stable function reference. I've written about that here: [React Query Data Transformations](https://tkdodo.eu/blog/react-query-data-transformations#3-using-the-select-option).

One very strong argument for the current architecture is partial subscriptions. For example:

```
function useTodosQuery(select) {
  return useQuery("todos/all", loadTodos, { select });
}

function useTodosCountQuery() {
  return useTodosQuery(todosQuery, (todos) => todos.length);
}
```

`useTodosCountQuery` only subscribes you to the length of the todos. Which means that if you get a background update where some todos have gotten different names, but the length hasn't changed, all observers subscribed via `useTodosCountQuery` will _not_ get informed about those changes and thus not re-render.

Your proposal:

```
function useTodosCountQuery() {
  const todosQuery = useTodosQuery();
  return useQuerySelector(todosQuery, (todos) => todos.length);
}
```

cannot do that because the observer will get informed about the updated todos via `useTodosQuery`, before running the selector and realizing that there are no changes to the length. But at that point, you cannot bail out of the rendering anymore.

**Reply by zaverden** [reply]
thank you for the reply. got your point

**Reply by Klemminn** [reply]
How would you go about typing select in such a use case, with the intention of inferring the correct return type?

`function useTodosQuery(select: ?) {...`

**Reply by TkDodo** [reply]
```
type Todos = Array<Todo>

const loadTodos = (): Promise<Todos> => ...

const useTodosQuery = <TResult = Todos>(select: (data: Todos => TResult)) {
  return useQuery(["todos", "all"], loadTodos, { select });
}
```

**Reply by Klemminn** [reply]
How are you so nice and quick to respond? 🤩

Just want to say that I've been through this repo a lot in the last few months, and I'm always more and more amazed about how much time and effort you put into answering everything. Even not-that related questions like mine, which I do realize was more of a problem with my amateur TypeScript abilities than react-query.

Thank you for being you! We are very lucky to have you 🥰

**Reply by gromaco** [reply]
> type Todos = Array<Todo>
> 
> const loadTodos = (): Promise<Todos> => ...
> 
> const useTodosQuery = <TResult = Todos>(select: (data: Todos => TResult)) {
>   return useQuery(["todos", "all"], loadTodos, { select });
> }

@TkDodo it would be awesome to add this to your blog as an example for `select`(maybe I missed it) - it could save a lot of time for someone like me 😃

**Comment by patrickherrmann-productplan** [comment]
Sorry if it's not good form to tack onto an exiting question. Why is `select` not a method on `UseQueryResult` rather than an option for `useQuery`? This would make it much easier to define variations derived from the same cached query, for example:

```ts
function useTodosQuery() {
  return useQuery("todos/all", loadTodos); // Returns UseQueryResult<Todo[]>
}

function useTodosCountQuery() {
  return useTodosQuery().select(todos => todos.length); // Returns UseQueryResult<number>
}
```

It would also empower calling components to perform context-specific transformations, and probably solve the type inference issues `select` seems to have.

**Reply by TkDodo** [reply]
> Why is select not a method on UseQueryResult rather than an option for useQuery?

I think it wouldn't allow us to optimize re-renders as much as it does now, at least not with the way how subscriptions are done now. I guess it's the same as why redux isn't:

```
const store = useStore()
const todos = store.select(state => state.todos)
```

> and probably solve the type inference issues select seems to have.

`select` doesn't have type inference issues. If you provide generics to `useQuery`, you ruin type-inference, and there is no partial type inference (yet) in TypeScript so you basically "opt out" of it.
